python/ray/serve/_private/metrics_utils.py

```diff
@@ -165,52 +178,205 @@ def _get_datapoints(
-              from this timestamp until now.
-            do_compact: whether or not to delete the datapoints that's
-              before `window_start_timestamp_s` to save memory. Default is
-              true.
+            keys: Iterable of keys to aggregate across.
         Returns:
-            Max value of the data points for the key on and after time
-            window_start_timestamp_s, or None if there are no such points.
+            A tuple of (float, int) where the first element is the sum across
+            all values found at `keys`, and the second is the number of valid
+            keys used to compute the sum.
+            Returns (None, 0) if no valid keys have data.
         """
-        points_after_idx = self._get_datapoints(key, window_start_timestamp_s)
+        return self._aggregate_reduce(keys, sum)
 
-        if do_compact:
-            self.data[key] = points_after_idx
+    def aggregate_avg(
+        self,
+        keys: Iterable[Hashable],
+    ) -> Tuple[Optional[float], int]:
+        """Average the entire set of timeseries values across the specified keys.
 
-        return max((point.value for point in points_after_idx), default=None)
+        Args:
+            keys: Iterable of keys to aggregate across.
+        Returns:
+            A tuple of (float, int) where the first element is the mean across
+            all values found at `keys`, and the second is the number of valid
+            keys used to compute the mean.
+            Returns (None, 0) if no valid keys have data.
+        """
+        return self._aggregate_reduce(keys, statistics.mean)
+
+
+def _bucket_latest_by_window(
+    series: List[TimeStampedValue],
+    start: float,
+    window_s: float,
+) -> Dict[int, float]:
+    """
+    Map each window index -> latest value seen in that window.
+    Assumes series is sorted by timestamp ascending.
+    """
+    buckets: Dict[int, float] = {}
+    for p in series:
+        w = int((p.timestamp - start) // window_s)
+        buckets[w] = p.value  # overwrite keeps the latest within the window
+    return buckets
+
+
+def _merge_two_timeseries(
+    t1: List[TimeStampedValue], t2: List[TimeStampedValue], window_s: float
+) -> List[TimeStampedValue]:
+    """
+    Merge two ascending time series by summing values within a specified time window.
+    If multiple values fall within the same window in a series, the latest value is used.
+    The output contains one point per window that had at least one value, timestamped
+    at the window center.
+    """
+    if window_s <= 0:
+        raise ValueError(f"window_s must be positive, got {window_s}")
+
+    if not t1 and not t2:
+        return []
+
+    # Align windows so each output timestamp sits at the center of its window.
+    # start is snapped to window_s boundary for binning stability
+    earliest = min(x[0].timestamp for x in (t1, t2) if x)
+    start = earliest // window_s * window_s - (window_s / 2.0)
```

Comment:
Let's just use the start of the window instead of the center, I think it is clearer

```suggestion
    start = earliest // window_s * window_s
```